src/main/java/backend/Logic.java

```diff
@@ -247,13 +209,204 @@ public Model getRepo(String repoId) {
+     * from the repository source.
+     *
+     * @param filterTexts Filter expressions to process
+     */
+    public void filterSortRefresh(List<FilterExpression> filterTexts) {
+        // Open specified repos
+        openRepositoriesInFilters(filterTexts);
+
+        // First filter, for issues requiring a metadata update.
+        Map<String, List<TurboIssue>> toUpdate = tallyMetadataUpdate(filterTexts);
+
+        if (toUpdate.size() > 0) {
+            // If there are issues requiring metadata update, we dispatch the metadata requests...
+            ArrayList<CompletableFuture<Boolean>> metadataRetrievalTasks = new ArrayList<>();
+            toUpdate.forEach((repoId, issues) -> {
+                metadataRetrievalTasks.add(getIssueMetadata(repoId, issues));
+            });
+            // ...and then wait for all of them to complete.
+            CompletableFuture<Void> allCompletables = CompletableFuture.allOf(
+                    metadataRetrievalTasks.toArray(new CompletableFuture[metadataRetrievalTasks.size()]));
+            allCompletables
+                    .thenCompose(n -> getRateLimitResetTime())
+                    .thenApply(this::updateRemainingRate)
+                    .thenRun(() -> updateUI(filterAndSort(filterTexts))); // Then filter the second time.
+        } else {
+            // If no issues requiring metadata update, just run the filter and sort.
+            updateUI(filterAndSort(filterTexts));
+        }
+    }
+
+    /**
+     * Given a list of filter expressions, open all repositories necessary for processing the filter expressions.
+     *
+     * @param filterTexts Filter expressions to process.
+     */
+    private void openRepositoriesInFilters(List<FilterExpression> filterTexts) {
+        HashSet<String> reposToOpen = new HashSet<>();
+
+        filterTexts.forEach(filterText -> {
+            filterText.find(Qualifier::isMetaQualifier).stream()
+                    .filter(metaQualifier ->
+                            metaQualifier.getName().equals(Qualifier.REPO) && metaQualifier.getContent().isPresent())
+                    .forEach(repoQualifier ->
+                            reposToOpen.add(repoQualifier.getContent().get()));
+        });
+
+        reposToOpen.forEach(this::openRepositoryFromFilter);
+    }
+
+    /**
+     * Given a list of filter expressions, determine issues within the model that require a metadata update.
+     *
+     * @param filterTexts Filter expressions to process for metadata requests.
+     * @return Repo IDs and the corresponding issues in the repo requiring a metadata update.
+     */
+    private Map<String, List<TurboIssue>> tallyMetadataUpdate(List<FilterExpression> filterTexts) {
+        HashMap<String, List<TurboIssue>> toUpdate = new HashMap<>();
+
+        List<TurboIssue> allModelIssues = models.getIssues();
+
+        filterTexts.stream().filter(Qualifier::hasUpdatedQualifier).forEach(filterText -> {
```

Comment:
Do we need to preprocess here to convert `updated-self` (and any other qualifier which requires metadata to work correctly) into `update`?